iocore/net/QUICNetVConnection.cc

```diff
@@ -591,10 +664,17 @@ QUICNetVConnection::state_connection_closing(int event, Event *data)
     break;
   case QUIC_EVENT_CLOSING_TIMEOUT:
     this->_close_closing_timeout(data);
+    break;
+
+  // active timeout signal VC_EVENT_IMMEDIATE
+  case VC_EVENT_IMMEDIATE:
+  case VC_EVENT_ACTIVE_TIMEOUT:
     this->_switch_to_close_state();
+    this->handleEvent(QUIC_EVENT_SHUTDOWN, nullptr);
```

Comment:
Why do we need to schedule this event and delay the cleanup? Is it possible to do the cleanup in _switch_to_close/draining_state()?

Reply:
Ah, I see. Originally `QUIC_EVENT_SHUTDOWN` was used for the final cleanup. That is the reason we introduced this event.

If we are going to only cleanup what we don't need anymore on the state, we don't need to delay the process, right?

Reply:
Yes we just pass QUIC_EVENT_SHUTDOWN to the close state's handler and free the qvc. and push it into closedQueue, since we are at the end of the closing/draining period.

Reply:
Maybe we misunderstood each other. Let me clarify my point.

As the spec says,

> Retaining packet protection keys is unnecessary once a
   connection is in the draining state.

we may free some resources on draining / closing state. So calling `_prepare_to_close_or_drain` on the states is fine. But I don't see a reason to use `QUIC_EVENT_SHUTDOWN` and `handleEvent` during the process. Why don't you call `_prepare_to_close_or_drain` In `switch_to_xxx_state()` ?


Also, we use `shecule_imm()` instead of calling an event handler directly to give a final chance to send a `CONNECTION_CLOSE` frame before releasing resources. But probably the event is not needed anymore because now we have closing timeout and its handler is the only caller of `switch_to_close_state`.